File: lib/src/graph/runtime_graph_replay.dart
import '../replay/replay_graph.dart';
import 'runtime_graph.dart';

Map<String, dynamic> replayGraphLineage(RuntimeGraph graph) {
  final replayed = replayRuntimeGraph(graph);
  return {
    'replayed': replayed.toJson(),
    'lineage_hash': graphReplayHash(replayed),
    'bounded': true,
  };
}

RuntimeGraph mergeGraphReplay(RuntimeGraph base, RuntimeGraph overlay) {
  final b = normalizeRuntimeGraph(base);
  final o = normalizeRuntimeGraph(overlay);
  return normalizeRuntimeGraph(RuntimeGraph(
    nodes: [...b.nodes, ...o.nodes],
    edges: [...b.edges, ...o.edges],
  ));
}

File: src/graph/runtimeGraphReplay.ts
import { RuntimeGraphContract, type RuntimeGraph } from "../contracts/graphContracts.js";
import { graphReplayHash, replayRuntimeGraph } from "../replay/replayGraph.js";

export function replayGraphLineage(graph: RuntimeGraph): {
  replayed: RuntimeGraph;
  lineage_hash: string;
  bounded: boolean;
} {
  const replayed = replayRuntimeGraph(graph);
  return {
    replayed,
    lineage_hash: graphReplayHash(replayed),
    bounded: true,
  };
}

export function mergeGraphReplay(base: RuntimeGraph, overlay: RuntimeGraph): RuntimeGraph {
  const b = RuntimeGraphContract.normalize(base);
  const o = RuntimeGraphContract.normalize(overlay);
  return RuntimeGraphContract.normalize({
    nodes: [...b.nodes, ...o.nodes],
    edges: [...b.edges, ...o.edges],
    bounded: true,
  });
}
